indexer/sql/src/non_blocking_indexer.rs

```diff
@@ -0,0 +1,343 @@
+            Some(block_to_index) => block_to_index,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        Ok(block_to_index.clone())
+    }
+
+    fn remove_or_fail(
+        blocks: Arc<Mutex<HashMap<u64, BlockToIndex>>>,
+        block_height: &u64,
+    ) -> error::Result<BlockToIndex> {
+        let mut blocks = blocks.lock().expect("Can't lock blocks");
+        let block_to_index = match blocks.remove_entry(block_height) {
+            Some(block) => block,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        #[cfg(feature = "tracing")]
+        tracing::debug!(
+            block_height = block_height,
+            blocks_len = blocks.len(),
+            "remove_or_fail called"
+        );
+        Ok(block_to_index.1)
+    }
+}
+
+impl Indexer for NonBlockingIndexer {
+    type Error = crate::error::IndexerError;
+
+    fn start(&mut self) -> error::Result<()> {
+        let context = self.context.clone();
+        block_call(self.runtime.as_ref(), &self.handle, async move {
+            context.migrate_db().await
+        })?;
+
+        self.indexing = true;
+        Ok(())
+    }
+
+    fn shutdown(&mut self) -> error::Result<()> {
+        // Avoid running this twice when called manually and from `Drop`
+        if !self.indexing {
+            return Ok(());
+        }
+        self.indexing = false;
+
+        // NOTE: This is to allow the indexer to commit all db transactions since this is done
+        // async. It just loops quickly making sure no indexing blocks are remaining.
+        for _ in 0..10 {
+            let blocks = self.blocks.lock().expect("Can't lock blocks");
+            if blocks.is_empty() {
+                break;
+            }
+
+            drop(blocks);
+
+            sleep(Duration::from_millis(10));
+        }
+
+        #[cfg(feature = "tracing")]
+        {
+            let blocks = self.blocks.lock().expect("Can't lock blocks");
+            if !blocks.is_empty() {
+                tracing::warn!("Some blocks are still being indexed, maybe non_blocking_indexer `post_indexing` wasn't called by the main app?");
+            }
+        }
+
+        Ok(())
+    }
+
+    fn pre_indexing(&self, _block_height: u64) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+        Ok(())
+    }
+
+    fn index_block(&self, block: &BlockInfo, _block_outcome: &BlockOutcome) -> error::Result<()> {
+        assert!(self.indexing, "Can't index after shutdown");
+
+        #[cfg(feature = "tracing")]
+        tracing::info!(block_height = block.height, "index_block called");
+        self.find_or_create(block, |_block_to_index| {
+            #[cfg(feature = "tracing")]
+            tracing::info!(block_height = block.height, "index_block started/finished");
+            Ok(())
+        })
+    }
```

Comment:
If this function return a Result, is better to don't panic here. Returning a result is better cause it could be handled